### PR TITLE
fixed sudo checks

### PR DESCRIPTION
**Pull request by rcmdnk**
To check `sudo` command,
it is better to check `/bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}"` command instead of using `-v`,
because users can have limited sudo right and they can `sudo -v` test.

In addition **Select the Linuxbrew installation directory** prompts were not working before.
This PR changes to do:

* Check `/usr/bin/sudo -n /bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}`
* If not, ask user to select (Enter password for /home/linuxbrew, C-D for $HOME/, C-C to interrupt)
* If Password is entered, try `/usr/bin/sudo /bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}`
* If not, or C-D, use $HOME/

### CODE REVIEW

File: install [hunk 1]
```diff
@@ -253,13 +253,17 @@ unless mac?
...
+
+
 # Tests will fail if the prefix exists, but we don't have execution
 # permissions. Abort in this case.
 abort <<-EOABORT if File.directory?(HOMEBREW_PREFIX) && (!File.executable? HOMEBREW_PREFIX)
 The Homebrew prefix, #{HOMEBREW_PREFIX}, exists but is not searchable. If this is
 not intentional, please restore the default permissions and try running the
 installer again:
-    sudo chmod 775 #{HOMEBREW_PREFIX}
+    sudo chmod 775 $USER #{HOMEBREW_PREFIX}
```
Comment (sjackman) [inline install]: What's the purpose of this change?

**Comment by sjackman** [comment]
I'm happy with the purpose of this PR. Thanks for this work, @rcmdnk. The patch seems bigger than necessary though. Can we instead replace `/usr/bin/sudo -v` with `/usr/bin/sudo -l /bin/mkdir >/dev/null` ?

File: install [hunk 2]
```diff
@@ -218,10 +211,6 @@ def shell_profile
   end
 end
 
-# Invalidate sudo timestamp before exiting (if it wasn't active before).
-Kernel.system "/usr/bin/sudo -n -v 2>/dev/null"
-at_exit { Kernel.system "/usr/bin/sudo", "-k" } unless $CHILD_STATUS.success?
-
```
Comment (sjackman) [inline install]: Instead of moving this chunk let's patch line 222.

File: install [hunk 3]
```diff
@@ -218,10 +211,6 @@ def shell_profile
   end
 end
 
-# Invalidate sudo timestamp before exiting (if it wasn't active before).
-Kernel.system "/usr/bin/sudo -n -v 2>/dev/null"
```
Comment (sjackman) [inline install]: You can prevent the prompt by disabling this line on Linux.
```ruby
Kernel.system "/usr/bin/sudo -n -v 2>/dev/null" if mac?
```

File: install [hunk 4]
```diff
@@ -233,18 +222,29 @@ abort <<-EOABORT unless !mac? || `dsmemberutil checkmembership -U "#{ENV["USER"]
...
   if File.writable?(HOMEBREW_PREFIX_DEFAULT) || File.writable?("/home/linuxbrew") || File.writable?("/home")
     HOMEBREW_PREFIX = HOMEBREW_PREFIX_DEFAULT.freeze
   else
-    Kernel.system "/usr/bin/sudo -n -v 2>/dev/null"
-    unless $CHILD_STATUS.success?
+    io = IO.popen "/usr/bin/sudo -n /bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}", :err => [:child, :out]
+    out = io.read
+    io.close
+    @have_sudo = $CHILD_STATUS.success?
+    if (not @have_sudo) && out == "sudo: a password is required\n"
```
Comment (sjackman) [inline install]: How about…
```ruby
    Kernel.system "/usr/bin/sudo -n -l /bin/mkdir >/dev/null 2>/dev/null"
    unless $CHILD_STATUS.success?
…
```

File: install [hunk 5]
```diff
@@ -253,13 +253,17 @@ unless mac?
   HOMEBREW_REPOSITORY = "#{HOMEBREW_PREFIX}/Homebrew".freeze
 end
 
+# Invalidate sudo timestamp before exiting (if it wasn't active before).
+at_exit { Kernel.system "/usr/bin/sudo", "-k" } if @have_sudo
+
+
```
Comment (sjackman) [inline install]: Please remove this chunk.

File: install-ruby [hunk 6]
```diff
@@ -9,9 +9,8 @@ case $(uname -m) in
   *) echo "Error: Your CPU $(uname -m) is not supported." >&2; exit 1 ;;
 esac
 
-if test -w /home/linuxbrew || sudo -v; then
+if test -w /home/linuxbrew || sudo mkdir -p /home/linuxbrew; then
```
Comment (sjackman) [inline install-ruby]: I like this change. Thanks! Could you please add `sudo -p` to add a custom prompt?
```suggestion
if test -w /home/linuxbrew || /usr/bin/sudo -p "[sudo] Enter password for $USER to install Ruby: " mkdir -p /home/linuxbrew; then
```

**Review by sjackman — CHANGES_REQUESTED** [review]
Thanks for this work!

File: install (hunk 1 above)
Comment (rcmdnk) [reply install]: Sorry, this is just a misunderstanding that I saw somehow `chmod` as `chown`.
`$USER` was removed.

File: install-ruby (hunk 6 above)
Comment (rcmdnk) [reply install-ruby]: This seems good, included.

File: install (hunk 3 above)
Comment (rcmdnk) [reply install]: It is unnecessary because `sudo` will be checked only when it is needed in below.

File: install (hunk 2 above)
Comment (rcmdnk) [reply install]: As `sudo` is used only when it is needed, it should be added below.

File: install (hunk 4 above)
Comment (rcmdnk) [reply install]: Some users have right to do `sudo` with password, and some have right w/o password.
This is needed to distinguish users which do not have the right anyway.
Otherwise, the next prompt will be strange for these users.

File: install (hunk 5 above)
Comment (rcmdnk) [reply install]: This needs to be here because `have_sudo` is just set above.

File: install (hunk 4 above)
Comment (rcmdnk) [reply install]: `sudo -l /bin/mkdir` instead of `sudo /bin/mkdir -p #{HOMEBREW_PREFIX_DEFAULT}` seems fine.

File: install (hunk 3 above)
Comment (sjackman) [reply install]: Yep, so disable this line on Linux as above.

File: install (hunk 4 above)
Comment (sjackman) [reply install]: You can use `sudo -nv` to check whether a user has `sudo` access without a password.

File: install (hunk 5 above)
Comment (sjackman) [reply install]: I believe `sudo -k` is successful whether or not the user has `sudo` access, so it doesn't need to be conditional on `@have_sudo`.

File: install (hunk 2 above)
Comment (sjackman) [reply install]: Let's solve one issue per PR. This issue need not be address in this PR.

File: install (hunk 3 above)
Comment (rcmdnk) [reply install]: On Mac, in where `sudo` requires a password,
`sudo -n -v` fails.
Therefore these lines are meaningless `if mac` is added.

File: install (hunk 2 above)
Comment (rcmdnk) [reply install]: This is directly related to `sudo` check below.
If a password is required, this simple check fails and does not implement `sudo -k` at exit.

File: install (hunk 4 above)
Comment (rcmdnk) [reply install]: Yes, but it fails if the user can do `sudo` only with a password.
Therefore, here, 

1. first check if the user can use `sudo` without password by `sudo -n -l mkdir`.
2. If failed, check if a password is required or not by the output.
3. Only if a password is required, ask user to put password to create /home/linuxbrew otherwise make $HOME/.linuxbrew

File: install (hunk 3 above)
Comment (sjackman) [reply install]: These lines are from Homebrew/install on macOS.
See https://github.com/Homebrew/install/blob/8e6decdff581a4a463d66c363831f8da6b865ff8/install#L148-L150

File: install (hunk 4 above)
Comment (sjackman) [reply install]: I suggest…
```ruby
if `sudo -nv 2>&1` == "sudo: a password is required\n"
```

File: install (hunk 5 above)
Comment (rcmdnk) [reply install]: Yes, both cases work in the same as a result.

I just realized that if we want to implement correctly as a comment **(if it wasn't active before)**,
we put this line after line 242.

File: install (hunk 3 above)
Comment (rcmdnk) [reply install]: Sorry, the previous comment is wrong and I had misunderstandings.
This `at_exit` code works as expected and do not needed to be moved,
as `sudo -k` do nothing as you mentioned.

I will change them back as before about `at_exit`.

File: install (hunk 4 above)
Comment (rcmdnk) [reply install]: Yes, that is enough in this case. updated.

**Comment by sjackman** [comment]
Please fix these style errors:
```
$ gem install rubocop
$ rubocop
Inspecting 3 files
C..
Offenses:
install:236:9: C: Style/Not: Use ! instead of not.
    if (not @have_sudo) && out == "sudo: a password is required\n"
        ^^^
install:244:7: C: Metrics/BlockNesting: Avoid more than 3 levels of block nesting.
      rescue Interrupt ...
      ^^^^^^^^^^^^^^^^
3 files inspected, 2 offenses detected
The command "rubocop" exited with 1.
```
See the TravisCI output https://travis-ci.org/Linuxbrew/install/builds/490450955

**Comment by rcmdnk** [comment]
To reduce block nests, a new function `prefix` was implemented.

File: install [hunk 7]
```diff
@@ -91,6 +90,37 @@ def system(*args)
...
+      exit!
+    end
+  end
+  if @have_sudo
+    return prefix_default
+  else
+    return "#{ENV["HOME"]}/.linuxbrew"
+  end
+end
+
```
Comment (sjackman) [inline install:102]: Please avoid large diff changes.

Comment (sjackman) [reply install:102]: They are difficult to review.

**Comment by rcmdnk** [comment]
I refactored the code and commits.
I think this is a minimum change to have similar functions as before.

File: install [hunk 8]
```diff
@@ -94,8 +94,12 @@ end
 def sudo?
   return @have_sudo unless @have_sudo.nil?
 
-  Kernel.system "/usr/bin/sudo", "-v"
-  @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
+  begin
+    Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+    @have_sudo = $CHILD_STATUS.success?
```
Comment (sjackman) [inline install]:
```suggestion
    @have_sudo = $CHILD_STATUS&.success?
```
`$CHILD_STATUS` can be `nil` if `/usr/bin/sudo` does not exist.

File: install [hunk 9]
```diff
@@ -94,8 +94,12 @@ end
...
+  rescue Interrupt
+    exit!
+  end
```
Comment (sjackman) [inline install]: Is the `begin/rescue/end` block needed? Please remove it if not.

File: install [hunk 10]
```diff
@@ -237,8 +241,8 @@ unless mac?
   if File.writable?(HOMEBREW_PREFIX_DEFAULT) || File.writable?("/home/linuxbrew") || File.writable?("/home")
     HOMEBREW_PREFIX = HOMEBREW_PREFIX_DEFAULT.freeze
   else
-    Kernel.system "/usr/bin/sudo -n -v 2>/dev/null"
-    unless $CHILD_STATUS.success?
+    out = `/usr/bin/sudo -n -l mkdir 2>&1`
+    if !$CHILD_STATUS.success? && out == "sudo: a password is required\n"
```
Comment (sjackman) [inline install]: Could you please rename `out` to `sudo_output`?

**Comment by sjackman** [comment]
This PR is getting much closer. Thank you, @rcmdnk!

**Comment by sjackman** [comment]
If you know how to squash a PR down to a single commit using `git rebase -i origin/master`, please do that. If you have no idea what I'm talking about, don't worry about it. I'll squash it when I merge it.

**Comment by rcmdnk** [comment]
Yes, these commits were squished.

**Comment by sjackman** [comment]
@rcmdnk Are you interested in continuing with this PR?

**Comment by rcmdnk** [comment]
Yes, it should be merged otherwise the script does not work in some of my nodes.

I squished commits in a single commit.
Do you need any other changes to merge?

**Comment by sjackman** [comment]
Yes, there's a number of unresolved comments.
See https://github.com/Linuxbrew/install/pull/47/files

File: install (hunk 8 above)
Comment (rcmdnk) [reply install]: Changed to 

    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success? || false

to keep compatibility for Ruby 2.2

File: install (hunk 9 above)
Comment (rcmdnk) [reply install]: If this does not exist,
Ctrl-C shows traceback message.

File: install (hunk 10 above)
Comment (rcmdnk) [reply install]: Renamed

**Comment by rcmdnk** [comment]
Sorry, I missed these comments.
I updated a commit for them.

File: install (hunk 9 above)
Comment (sjackman) [reply install]: Thanks for the explanation. I believe `exit` should work and also prevent the traceback. Let's use that rather than `exit!`.

File: install (hunk 8 above)
Comment (sjackman) [reply install]: No need for the `|| false`. The falsey value of `nil` for false is fine.

File: install [hunk 11]
```diff
@@ -94,8 +94,12 @@ end
 def sudo?
   return @have_sudo unless @have_sudo.nil?
 
-  Kernel.system "/usr/bin/sudo", "-v"
-  @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
+  begin
+    Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success? || false
+  rescue Interrupt
+    exit!
```
Comment (sjackman) [inline install]:
```suggestion
    exit
```
`exit` also prevents the traceback.

File: install [hunk 12]
```diff
@@ -94,8 +94,12 @@ end
 def sudo?
   return @have_sudo unless @have_sudo.nil?
 
-  Kernel.system "/usr/bin/sudo", "-v"
-  @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
+  begin
+    Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success? || false
```
Comment (sjackman) [inline install]:
```suggestion
    @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
```
`nil` for `false` is fine.

Comment (sjackman) [reply install]:
```suggestion
    @have_sudo = $CHILD_STATUS.success?
```
I was wrong here. `$CHILD_STATUS` is not `nil` when `/usr/bin/sudo` does not exist.

Comment (rcmdnk) [reply install]: Maybe I also did a similar misunderstanding that I observed `nil` before adding `require "English" in the test script...
The last one is enough, I agree.

I put `|| false` because if `@have_sudo` is nil, `sudo?` cannot pass the first line always.

Anyways, I updated to just `$CHILD_STATUS.success?`.

File: install (hunk 11 above)
Comment (rcmdnk) [reply install]: As `SystemExit` trap does not exist, they work almost the same.
But maybe `exit` is better for the future update.

The difference here is just a default exit code (1 or 0).
I thought it should be an error exit, but you can decide later if necessary.

I changed it to `exit`.

File: install (hunk 9 above)
Comment (rcmdnk) [reply install]: I think I misunderstood your comment.
`begin`~`end` is not necessary here, so I removed it.

File: install (hunk 12 above)
Comment (sjackman) [reply install]: > Maybe I also did a similar misunderstanding that I observed nil before adding `require "English" in the test script...

I think I made the identical error.

> I put || false because if @have_sudo is nil, sudo? cannot pass the first line always.

Yes, good point. I missed that.

File: install (hunk 11 above)
Comment (sjackman) [reply install]: Good point. `exit 1` is probably better.
```suggestion
    exit 1
```

File: install [hunk 13]
```diff
@@ -94,8 +94,10 @@ end
 def sudo?
   return @have_sudo unless @have_sudo.nil?
 
-  Kernel.system "/usr/bin/sudo", "-v"
-  @have_sudo = $CHILD_STATUS && $CHILD_STATUS.success?
+  Kernel.system "/usr/bin/sudo", "-l", "mkdir"
+  @have_sudo = $CHILD_STATUS.success?
+rescue Interrupt
+  exit
```
Comment (sjackman) [inline install:100]:
```suggestion
  exit 1
```

**Review by sjackman — APPROVED** [review]

**Comment by sjackman** [comment]
Thank you very much for these improvements, @rcmdnk ! And thank you for your first contribution to Linuxbrew/install! 🏅
